### PR TITLE
Allow opting out of `.flutter-plugins`, opt-out in `refreshPluginsList`.

### PR DESCRIPTION
**Pull request by matanlurey**
Work towards https://github.com/flutter/flutter/issues/48918.

Workaround for https://github.com/flutter/flutter/issues/157391 (see https://github.com/flutter/flutter/pull/157393#issuecomment-2434336047).

I'll run all post-submit tasks as well on this PR using `test: all`.

### CODE REVIEW

**Comment by matanlurey** [comment]
Ah this needs another patch first (https://github.com/flutter/flutter/pull/157393/files#diff-68361f6fc9aafa5fa7de042eec4603dd4f321eb783a00edd57780ef78c1094bb).

Moving to draft for a moment.

**Comment by matanlurey** [comment]
Sorry for the noise, PTAL.

**Review by jmagman — APPROVED** [review]
Thanks for figuring out a minimal workaround.

File: packages/flutter_tools/test/general.shard/plugins_test.dart [hunk 1]
```diff
@@ -448,6 +448,20 @@ dependencies:
...
+          'plugin_d',
+          'plugin_a',
+          '/local_plugins/plugin_c',
+          '/local_plugins/plugin_b',
+        ]);
+
+        await refreshPluginsList(flutterProject, writeLegacyPluginsList: false);
+
+        expect(flutterProject.flutterPluginsFile.existsSync(), false);
+        expect(flutterProject.flutterPluginsDependenciesFile.existsSync(), true);
```
Comment (andrewkolos) [inline packages/flutter_tools/test/general.shard/plugins_test.dart:462]: Nit
```suggestion
        expect(flutterProject.flutterPluginsFile, exists);
        expect(flutterProject.flutterPluginsDependenciesFile, isNot(exists));
```

File: packages/flutter_tools/lib/src/macos/cocoapod_utils.dart [hunk 2]
```diff
@@ -31,6 +31,12 @@ Future<void> processPodsIfNeeded(
     iosPlatform: project.ios.existsSync(),
     macOSPlatform: project.macos.existsSync(),
     forceCocoaPodsOnly: forceCocoaPodsOnly,
+    // TODO(matanlurey): As-per discussion on https://github.com/flutter/flutter/pull/157393
+    // we'll assume that iOS/MacOS builds do not use or rely on the `.flutter-plugins` legacy
+    // file being generated. A better long-term fix would be not to have a call to refreshPluginsList
+    // at all, and instead have it implicitly run by the FlutterCommand instead. See
+    // https://github.com/flutter/flutter/issues/157391 for details.
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/macos/cocoapod_utils.dart:38]: Turbo-super-nit:

VSCode will parse all of this as a single todo if you add a space at the start of the proceeding lines:

```suggestion
    // TODO(matanlurey): As-per discussion on https://github.com/flutter/flutter/pull/157393
    //  we'll assume that iOS/MacOS builds do not use or rely on the `.flutter-plugins` legacy
    //  file being generated. A better long-term fix would be not to have a call to refreshPluginsList
    //  at all, and instead have it implicitly run by the FlutterCommand instead. See
    //  https://github.com/flutter/flutter/issues/157391 for details.
```

Then again, I don't know if I actually want more squiggly lines in my editor 🤷

**Review by andrewkolos — APPROVED** [review]
LGTM

**Comment by matanlurey** [comment]
It looks like this is unrelated.

**Pull request merged**